src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/histogram.js

```diff
@@ -0,0 +1,683 @@
+    const axis = this.axisSpec();
+    const canvas = this.canvas;
+
+    const fontStyle = "10px DejaVuSansMonoBook";
+    if (axis.x.label !== undefined) {
+      let padding_y = 20;
+      this.xAxisLabel
+        .attr("y", canvas.inner.height + canvas.margin.top + padding_y)
+        .text(axis.x.label)
+        .attr(
+          "x",
+          canvas.margin.left + this.textWidth(axis.x.label, fontStyle) / 2
+        );
+    }
+
+    if (axis.y.label !== undefined) {
+      const padding_y = 15;
+      this.yAxisLabel
+        .attr("y", -canvas.margin.left + padding_y)
+        .attr(
+          "x",
+          -canvas.margin.top -
+            canvas.inner.height / 2 +
+            this.textWidth(axis.y.label, fontStyle) / 2
+        )
+        .text(axis.y.label);
+    }
+  }
+
+  /**
+   * Return the text width. Ensures that labels on x axis wont overlap to keeps them readable.
+   */
+  textWidth(text, font) {
+    const canvas = document.createElement("canvas");
+    const context = canvas.getContext("2d");
+    context.font = font;
+    const metrics = context.measureText("  " + text);
+    return metrics.width;
+  }
+
+  /**
+   * Return the extrema of the data and and paddings that ensure data will fit into the
+   * drawing area.
+   *
+   * It traverses through data getting minimal and maximal values, and calculates padding based on
+   * span calculated from above values, multiplied by 10% so that the plot is a little bit smaller
+   * than the container.
+   */
+  extremesAndDeltas() {
+    const dataPoints = this.data();
+    let xMin = dataPoints[0];
+    let xMax = dataPoints[0];
+
+    dataPoints.forEach((value) => {
+      if (value < xMin) {
+        xMin = value;
+      }
+      if (value > xMax) {
+        xMax = value;
+      }
+    });
+
+    const dx = xMax - xMin;
+    const paddingX = 0.1 * dx;
+
+    return { xMin, xMax, paddingX, dx };
+  }
+
+  /**
+   * Return margins for plots drawing area.
+   */
+  margins() {
+    const axis = this.axisSpec();
+    const noXAxis = axis.x.label === undefined;
+    const noYAxis = axis.y.label === undefined;
+
+    if (noXAxis && noYAxis) {
+      return { top: 20, right: 20, bottom: 25, left: 35 };
+    } else if (noYAxis) {
+      return { top: 10, right: 20, bottom: 35, left: 35 };
+    } else if (noXAxis) {
+      return { top: 20, right: 10, bottom: 25, left: 55 };
+    }
+    return { top: 10, right: 10, bottom: 35, left: 55 };
+  }
+
+  /**
+   * Creates HTML div element as container for plot.
+   */
+  createOuterContainerWithStyle(width, height) {
+    const divElem = document.createElementNS(null, "div");
+    divElem.setAttributeNS(null, "class", "vis-histogram");
+    divElem.setAttributeNS(
+      null,
+      "viewBox",
+      0 + " " + 0 + " " + width + " " + height
+    );
+    divElem.setAttributeNS(null, "width", "100%");
+    divElem.setAttributeNS(null, "height", "100%");
+    divElem.setAttributeNS(null, "transform", "matrix(1 0 0 -1 0 0)");
```

Comment:
magical mmatrix hardcoded. Its purpose would probably not be obvious to the reader